storage: Maintain a separate set of unquiesced replicas

bdarnell opened the pull request:
This means that idle replicas no longer have a per-tick CPU cost,
which is one of the bottlenecks limiting the amount of data we can
handle per store.

Fixes #17609

Release note (performance improvement): Reduced CPU overhead of idle
ranges


The first five commits are from #24920; that PR should be merged and tested in isolation first.

bdarnell committed:
storage: Remove COCKROACH_ENABLE_TICK_QUIESCED

This is no longer used

Release note: None

bdarnell committed:
storage: Maintain a separate set of unquiesced replicas

This means that idle replicas no longer have a per-tick CPU cost,
which is one of the bottlenecks limiting the amount of data we can
handle per store.

Fixes #17609

Release note (performance improvement): Reduced CPU overhead of idle
ranges

bdarnell committed:
syncutil: Document that nil values are not supported.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24956)
<!-- Reviewable:end -->

petermattis commented:
Reviewed 2 of 2 files at r1, 4 of 4 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: 2 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/storage/replica.go, line 592 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnPbgvEYRPGx8K-k3y:-LAnPbgvEYRPGx8K-k3z:b-u8rgqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4295d2a2fdec654824cf98cdb420dbb62b07165d/pkg/storage/replica.go#L592)):*
> ```Go
> 	anotherOwnsLease := r.leaseStatus(*r.mu.state.Lease, r.store.Clock().Now(), r.mu.minLeaseProposedTS).State ==
> 		LeaseState_VALID &&
> 		!r.mu.state.Lease.OwnedBy(r.store.StoreID())
> ```

We've historically had subtle errors in logic like this. Perhaps this code should be restructured for testability. For example, the logic could be placed in a `shouldCampaignOnWakeLocked` function which takes a lease and other information and returns a boolean.

---

*[pkg/storage/replica.go, line 3914 at r8](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnSyb92Dg_zTefCovd:-LAnSyb92Dg_zTefCove:b3m1fg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/646debb8d610ff17b88410bc32d292c8513bb9e5/pkg/storage/replica.go#L3914)):*
> ```Go
> 
> // needsTick returns true if tick() should be called on this replica.
> func (r *Replica) needsTick() bool {
> ```

Is this method being used anymore?

---

*[pkg/storage/store.go, line 511 at r8](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnQyFv4HA0ri1rWIgs:-LAnQyFwDRbX33iSUMjd:b-z6ovad) ([raw file](https://github.com/cockroachdb/cockroach/blob/646debb8d610ff17b88410bc32d292c8513bb9e5/pkg/storage/store.go#L511)):*
> ```Go
> 	}
> 
> 	// The uniquesed subset of mu.replicas.
> ```

`s/uniquesed/unquiesced/g`

---

*[pkg/storage/store.go, line 512 at r8](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnR0-Q3pzmxjZuAw79:-LAnR0-Q3pzmxjZuAw7A:bmkz1zx) ([raw file](https://github.com/cockroachdb/cockroach/blob/646debb8d610ff17b88410bc32d292c8513bb9e5/pkg/storage/store.go#L512)):*
> ```Go
> 
> 	// The uniquesed subset of mu.replicas.
> 	unquiescedReplicas syncutil.IntMap
> ```

`syncutil.IntMap` might not be a good idea here as it is optimized for read-mostly workloads. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24956)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod @petermattis' comments and waiting until the precursor has merged.

---

Reviewed 2 of 4 files at r2, 2 of 2 files at r6, 1 of 1 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 2308 at r8](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnf-02AOIPrbLXEY6g:-LAnf-02AOIPrbLXEY6h:b-h9l0yi) ([raw file](https://github.com/cockroachdb/cockroach/blob/646debb8d610ff17b88410bc32d292c8513bb9e5/pkg/storage/store.go#L2308)):*
> ```Go
> 		return errors.Errorf("%s: replica already exists", repl)
> 	}
> 	s.unquiescedReplicas.Store(int64(repl.RangeID), unsafe.Pointer(repl))
> ```

`LoadOrStore` for symmetry (and an assertion)?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24956#-:-LAnfDLW6JP-eivDjUKo:b-gyq510)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/storage/replica.go, line 3914 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnSyb92Dg_zTefCovd:-LCZDoN80kfVIW0bTInn:b-pyrbqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/646debb8d610ff17b88410bc32d292c8513bb9e5/pkg/storage/replica.go#L3914)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this method being used anymore?
</blockquote></details>

No. Removed.

---

*[pkg/storage/store.go, line 511 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnQyFv4HA0ri1rWIgs:-LCZDvUY9SILJcv5MBJW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/646debb8d610ff17b88410bc32d292c8513bb9e5/pkg/storage/store.go#L511)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/uniquesed/unquiesced/g`
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 512 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnR0-Q3pzmxjZuAw79:-LCZE9eQE0RIqC-Ucp0R:b28k4i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/646debb8d610ff17b88410bc32d292c8513bb9e5/pkg/storage/store.go#L512)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`syncutil.IntMap` might not be a good idea here as it is optimized for read-mostly workloads. 
</blockquote></details>

OK, changed to a regular map with a mutex (a new mutex, since lock ordering gets tricky if we add it to `Store.mu`)

---

*[pkg/storage/store.go, line 2308 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnf-02AOIPrbLXEY6g:-LCZEeJu07tSLoFAeoAd:b9n5j31) ([raw file](https://github.com/cockroachdb/cockroach/blob/646debb8d610ff17b88410bc32d292c8513bb9e5/pkg/storage/store.go#L2308)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`LoadOrStore` for symmetry (and an assertion)?
</blockquote></details>

This has been changed to a regular map.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo unaddressed comment.

Also, did you do any testing that shows that this reduces the CPU overhead of idle ranges? I believe I filed this issue when doing some scalability testing where I artificially created 500k ranges on a cluster. See the `kv/splits/nodes=3` roachtest. 

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 592 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnPbgvEYRPGx8K-k3y:-LCZtboo-nC1EdnRaZiQ:b-82oqa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4295d2a2fdec654824cf98cdb420dbb62b07165d/pkg/storage/replica.go#L592)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We've historically had subtle errors in logic like this. Perhaps this code should be restructured for testability. For example, the logic could be placed in a `shouldCampaignOnWakeLocked` function which takes a lease and other information and returns a boolean.
</blockquote></details>

Ping.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24956#-:-LCZtI82BHqFjWUUN2i8:b86m48o)*
<!-- Sent from Reviewable.io -->

bdarnell committed:
storage: Refactor maybeCampaignOnWakeLocked for testability

Release note: None

bdarnell commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/replica.go, line 592 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/24956#-LAnPbgvEYRPGx8K-k3y:-LCZvJiZ4VoN0302AtbF:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4295d2a2fdec654824cf98cdb420dbb62b07165d/pkg/storage/replica.go#L592)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ping.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24956#-:-LD1vOUG0-4d11LLtfe5:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
bors r=petermattis

I have not done any testing with large numbers of ranges, FYI.
